pkg/tbtc/dkg.go

```diff
@@ -105,124 +113,214 @@ func (drl *dkgRetryLoop) start(
+// members. The excluded members list is built using the qualified operators
+// set. The algorithm that determines the qualified operators set does not
+// care about an exact mapping between operators and controlled members but
+// relies on the members count solely. That means the information about
+// readiness of specific members controlled by the given operators is not
+// included in the resulting qualified operators set. In order to properly
+// decide about inclusion or exclusion of specific members of a given
+// qualified operator, we must take the ready members list into account.
+func (drl *dkgRetryLoop) performMembersSelection(
+	readyMembersIndexes []group.MemberIndex,
+) ([]group.MemberIndex, error) {
+	qualifiedOperatorsSet, err := drl.qualifiedOperatorsSet(readyMembersIndexes)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get qualified operators: [%w]", err)
+	}
+
+	excludedMembersIndexes := make([]group.MemberIndex, 0)
+	for i, operator := range drl.selectedOperators {
+		memberIndex := group.MemberIndex(i + 1)
 
-		// If this attempt pushes us below the group quorum we are falling
-		// back to the random retry algorithm that excludes specific members
-		// from the original group selection result returned by the sortition
-		// pool.
-		if len(qualifiedOperators) >= drl.chainConfig.GroupQuorum {
-			return qualifiedOperators.Set(), nil
+		included := qualifiedOperatorsSet[operator] &&
+			slices.Contains(readyMembersIndexes, memberIndex)
+
+		if !included {
+			excludedMembersIndexes = append(
+				excludedMembersIndexes,
+				memberIndex,
+			)
 		}
 	}
 
-	// In any other case, try to make a random retry.
+	return excludedMembersIndexes, nil
+}
+
+// qualifiedOperatorsSet returns a set of operators qualified to participate
+// in the given DKG attempt. The set of qualified operators is taken from the
+// set of active operators who announced readiness through their controlled DKG
+// group members.
+func (drl *dkgRetryLoop) qualifiedOperatorsSet(
+	readyMembersIndexes []group.MemberIndex,
+) (map[chain.Address]bool, error) {
+	var readyOperators chain.Addresses
+	for _, memberIndex := range readyMembersIndexes {
+		readyOperators = append(
+			readyOperators,
+			drl.selectedOperators[memberIndex-1],
+		)
+	}
+
+	// For the first attempt, just return the operators who announced readiness.
+	// Otherwise, randomly exclude operators from the ready operators set.
+	if drl.attemptCounter == 1 {
+		return readyOperators.Set(), nil
+	}
+
+	// The retry algorithm expects that we count retries from 0. Since
+	// the first invocation of the algorithm will be for `attemptCounter == 1`
+	// we need to subtract one while determining the number of the given retry.
+	retryCount := drl.attemptCounter - 1
+
 	qualifiedOperators, err := retry.EvaluateRetryParticipantsForKeyGeneration(
-		drl.selectedOperators,
-		drl.randomRetrySeed,
-		drl.randomRetryCounter,
+		readyOperators,
+		drl.attemptSeed,
```

Comment:
Hmm, I don't think so.  We need the SHA-256 of the seed for the random retry and the plain-text seed for the announcer. The former needs a hash because we need to have a constant byte size. The latte uses plain text because I think this is safer for the future - if we need to extract the seed from the announcement for whatever reason, we can do it easily. Having two separate fields in the loop struct is a really low price for that.

Reply:
True, it is low price but it is also confusing. `attemptCounter` changes, `attemptStartBlock` changes, and my understanding during the review was that `attemptSeed` also changes somehow.

On the other hand, I see now that the current approach is aligned with `signing.go` and `seed` is like a `message`. So what we could do is to explain them in a short comment:
```
type dkgRetryLoop struct {
  // (...)

  // The original seed for DKG.
  // Used for the announcement. It never changes.       
  seed *big.Int

  // (...)

  // 8-bytes seed obtained from the original seed.
  // Used for the random operator selection. It never changes.
  attemptSeed int64
}
```

What do you think?

Reply:
Sure, see 3408c4ce